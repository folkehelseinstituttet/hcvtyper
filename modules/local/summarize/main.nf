process SUMMARIZE {

    label 'process_medium'
    errorStrategy 'terminate'

    // Environment with R tidyverse and seqinr packages from the conda-forge channel. Created using seqera containers.
    // URLs:
    // Docker image: https://wave.seqera.io/view/builds/bd-3536dd50a17de0ab_1?_gl=1*16bm7ov*_gcl_au*MTkxMjgxNTMwMi4xNzUzNzczOTQz
    // Singularity image: https://wave.seqera.io/view/builds/bd-88101835c4571845_1?_gl=1*5trzpp*_gcl_au*MTkxMjgxNTMwMi4xNzUzNzczOTQz
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/2a/2a1764abd77b9638883a202b96952a48f46cb0ee6c4f65874b836b9455a674d1/data':
        'community.wave.seqera.io/library/r-gridextra_r-png_r-seqinr_r-tidyverse:3536dd50a17de0ab' }"

    // Set environment variables for summarize.R
    env "PIPELINE_NAME",      "${workflow.manifest.name}"
    env "PIPELINE_VERSION",   "${workflow.manifest.version}"                 // Primary version source
    env "PIPELINE_REVISION",  "${workflow.revision ?: ''}"                  // Secondary (branch/tag info)
    env "PIPELINE_COMMIT",    "${workflow.commitId ?: ''}"                  // Tertiary (commit info)

    input:
    val version
    val name
    path samplesheet
    val stringency_1
    val stringency_2
    path 'trimmed/'
    path 'kraken_classified/'
    path 'parsefirst_mapping/'
    path 'stats_withdup/'
    path 'stats_markdup/'
    path 'depth/'
    path 'blast/'
    path 'glue/'
    path 'id/'
    path 'variation/'

    output:
    path 'Summary.csv'      , emit: summary
    path '*mqc.csv'         , emit: mqc
    path '*png'             , emit: png
    path "versions.yml"     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''

    """
    summarize.R \\
        $samplesheet \\
        $stringency_1 \\
        $stringency_2 \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
        tidyverse: \$(Rscript -e "library(tidyverse); cat(as.character(packageVersion('tidyverse')))")
        seqinr: \$(Rscript -e "library(seqinr); cat(as.character(packageVersion('seqinr')))")
        gridExtra: \$(Rscript -e "library(gridExtra); cat(as.character(packageVersion('gridExtra')))")
        png: \$(Rscript -e "library(png); cat(as.character(packageVersion('png')))")
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    """
    # Safe echo (interpolated by Groovy)
    echo "${args}"

    # Create realistic Summary.csv with proper header and sample data
    cat > Summary.csv << 'EOF'
sampleName,total_raw_reads,total_trimmed_reads,total_classified_reads,total_mapped_reads,fraction_mapped_reads_vs_median,Major_genotype_mapping,Major_reference,Minor_genotype_mapping,Minor_reference,major_typable,minor_typable,Reads_withdup_mapped_major,Reads_nodup_mapped_major,Percent_reads_mapped_of_trimmed_with_dups_major,Major_cov_breadth_min_5,Major_cov_breadth_min_10,percent_mapped_reads_major_firstmapping,Reads_withdup_mapped_minor,Reads_nodup_mapped_minor,Percent_reads_mapped_of_trimmed_with_dups_minor,Minor_cov_breadth_min_5,Minor_cov_breadth_min_10,percent_mapped_reads_minor_firstmapping,sequencer_id,Reads_nodup_mapped_first_mapping,Major_cov_breadth_min_1,Minor_cov_breadth_min_1,Major_avg_depth,Minor_avg_depth,Reference,GLUE_genotype,GLUE_subtype,glecaprevir,glecaprevir_mut,glecaprevir_mut_short,grazoprevir,grazoprevir_mut,grazoprevir_mut_short,paritaprevir,paritaprevir_mut,paritaprevir_mut_short,voxilaprevir,voxilaprevir_mut,voxilaprevir_mut_short,NS34A,NS34A_short,daclatasvir,daclatasvir_mut,daclatasvir_mut_short,elbasvir,elbasvir_mut,elbasvir_mut_short,ledipasvir,ledipasvir_mut,ledipasvir_mut_short,ombitasvir,ombitasvir_mut,ombitasvir_mut_short,pibrentasvir,pibrentasvir_mut,pibrentasvir_mut_short,velpatasvir,velpatasvir_mut,velpatasvir_mut_short,NS5A,NS5A_short,dasabuvir,dasabuvir_mut,dasabuvir_mut_short,sofosbuvir,sofosbuvir_mut,sofosbuvir_mut_short,NS5B,NS5B_short,HCV project version,GLUE engine version,PHE drug resistance extension version,script_name_stringency
Test_1,40000,34446,31482,31478,1.2804002521914215,NA,NA,NA,NA,NO,NO,NA,NA,NA,NA,NA,95.59,NA,NA,NA,NA,NA,3.55,@SRR24174266.1 1/1,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,HCVTyper (version unknown)
EOF

    # Create identical summary_mqc.csv (same data, MultiQC format)
    cat > summary_mqc.csv << 'EOF'
sampleName,total_raw_reads,total_trimmed_reads,total_classified_reads,total_mapped_reads,fraction_mapped_reads_vs_median,Major_genotype_mapping,Major_reference,Minor_genotype_mapping,Minor_reference,major_typable,minor_typable,Reads_withdup_mapped_major,Reads_nodup_mapped_major,Percent_reads_mapped_of_trimmed_with_dups_major,Major_cov_breadth_min_5,Major_cov_breadth_min_10,percent_mapped_reads_major_firstmapping,Reads_withdup_mapped_minor,Reads_nodup_mapped_minor,Percent_reads_mapped_of_trimmed_with_dups_minor,Minor_cov_breadth_min_5,Minor_cov_breadth_min_10,percent_mapped_reads_minor_firstmapping,sequencer_id,Reads_nodup_mapped_first_mapping,Major_cov_breadth_min_1,Minor_cov_breadth_min_1,Major_avg_depth,Minor_avg_depth,Reference,GLUE_genotype,GLUE_subtype,glecaprevir,glecaprevir_mut,glecaprevir_mut_short,grazoprevir,grazoprevir_mut,grazoprevir_mut_short,paritaprevir,paritaprevir_mut,paritaprevir_mut_short,voxilaprevir,voxilaprevir_mut,voxilaprevir_mut_short,NS34A,NS34A_short,daclatasvir,daclatasvir_mut,daclatasvir_mut_short,elbasvir,elbasvir_mut,elbasvir_mut_short,ledipasvir,ledipasvir_mut,ledipasvir_mut_short,ombitasvir,ombitasvir_mut,ombitasvir_mut_short,pibrentasvir,pibrentasvir_mut,pibrentasvir_mut_short,velpatasvir,velpatasvir_mut,velpatasvir_mut_short,NS5A,NS5A_short,dasabuvir,dasabuvir_mut,dasabuvir_mut_short,sofosbuvir,sofosbuvir_mut,sofosbuvir_mut_short,NS5B,NS5B_short,HCV project version,GLUE engine version,PHE drug resistance extension version,script_name_stringency
Test_1,40000,34446,31482,31478,1.2804002521914215,NA,NA,NA,NA,NO,NO,NA,NA,NA,NA,NA,95.59,NA,NA,NA,NA,NA,3.55,@SRR24174266.1 1/1,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,HCVTyper (version unknown)
EOF

    # Create a minimal PNG plot placeholder
    : > summary_plot.png

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
        tidyverse: \$(Rscript -e "library(tidyverse); cat(as.character(packageVersion('tidyverse')))")
        seqinr: \$(Rscript -e "library(seqinr); cat(as.character(packageVersion('seqinr')))")
        gridExtra: \$(Rscript -e "library(gridExtra); cat(as.character(packageVersion('gridExtra')))")
        png: \$(Rscript -e "library(png); cat(as.character(packageVersion('png')))")
    END_VERSIONS
    """
}
