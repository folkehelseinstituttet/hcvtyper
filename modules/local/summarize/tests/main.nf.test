nextflow_process {

    name "Test Process SUMMARIZE"
    script "../main.nf"
    process "SUMMARIZE"

    tag "modules"
    tag "summarize"

    test("summarize: comprehensive summary") {
        options "--profile docker"
        when {
            process {
                """
                // Process inputs - using file() instead of directory paths to properly stage contents
                input[0] = file("${projectDir}/modules/local/summarize/tests/data/samplesheet.csv", checkIfExists: true)
                input[1] = 80    // stringency_1
                input[2] = 95    // stringency_2

                // Stage directory contents using file() with glob patterns
                input[3] = file("${projectDir}/modules/local/summarize/tests/data/trimmed/*", checkIfExists: true)
                input[4] = file("${projectDir}/modules/local/summarize/tests/data/kraken_classified/*", checkIfExists: true)
                input[5] = file("${projectDir}/modules/local/summarize/tests/data/parsefirst_mapping/*", checkIfExists: true)
                input[6] = file("${projectDir}/modules/local/summarize/tests/data/stats_withdup/*", checkIfExists: true)
                input[7] = file("${projectDir}/modules/local/summarize/tests/data/stats_markdup/*", checkIfExists: true)
                input[8] = file("${projectDir}/modules/local/summarize/tests/data/depth/*", checkIfExists: true)
                input[9] = file("${projectDir}/modules/local/summarize/tests/data/blast/*", checkIfExists: true)
                input[10] = file("${projectDir}/modules/local/summarize/tests/data/glue/*", checkIfExists: true)
                input[11] = file("${projectDir}/modules/local/summarize/tests/data/id/*", checkIfExists: true)
                input[12] = file("${projectDir}/modules/local/summarize/tests/data/variation/*", checkIfExists: true)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot([
                    summary : process.out.summary,
                    mqc     : process.out.mqc,
                    png     : process.out.png,
                    versions: process.out.versions
                ]).match() }
            )
        }
    }

    test("summarize: stub") {
        options "-stub-run"
        when {
            process {
                """
                input[0] = file("${projectDir}/modules/local/summarize/tests/data/samplesheet.csv", checkIfExists: true)
                input[1] = 80
                input[2] = 95
                input[3] = file("${projectDir}/modules/local/summarize/tests/data/trimmed/*", checkIfExists: true)
                input[4] = file("${projectDir}/modules/local/summarize/tests/data/kraken_classified/*", checkIfExists: true)
                input[5] = file("${projectDir}/modules/local/summarize/tests/data/parsefirst_mapping/*", checkIfExists: true)
                input[6] = file("${projectDir}/modules/local/summarize/tests/data/stats_withdup/*", checkIfExists: true)
                input[7] = file("${projectDir}/modules/local/summarize/tests/data/stats_markdup/*", checkIfExists: true)
                input[8] = file("${projectDir}/modules/local/summarize/tests/data/depth/*", checkIfExists: true)
                input[9] = file("${projectDir}/modules/local/summarize/tests/data/blast/*", checkIfExists: true)
                input[10] = file("${projectDir}/modules/local/summarize/tests/data/glue/*", checkIfExists: true)
                input[11] = file("${projectDir}/modules/local/summarize/tests/data/id/*", checkIfExists: true)
                input[12] = file("${projectDir}/modules/local/summarize/tests/data/variation/*", checkIfExists: true)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot([
                    summary : process.out.summary,
                    mqc     : process.out.mqc,
                    png     : process.out.png,
                    versions: process.out.versions
                ]).match() }
            )
        }
    }
}
