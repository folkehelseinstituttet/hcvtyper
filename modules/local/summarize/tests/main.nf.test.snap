{
    "summarize: comprehensive summary": {
        "content": [
            {
                "summary": [
                    "Summary.csv:md5,5788053a8f5bf730235d9b5732713d97"
                ],
                "mqc": [
                    "summary_mqc.csv:md5,5788053a8f5bf730235d9b5732713d97"
                ],
                "png": [
                    "Variation_plot_major.png:md5,e19d8e3b61bdcb781ebb43e0328a6af4"
                ],
                "versions": [
                    "versions.yml:md5,bdc410e5c9e6ecec2ea6f41acaf3f8ef"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-08T12:13:25.270423308"
    },
    "summarize: stub": {
        "content": [
            {
                "summary": [
                    "Summary.csv:md5,093506819671498252da3ee2a6d3513a"
                ],
                "mqc": [
                    "summary_mqc.csv:md5,093506819671498252da3ee2a6d3513a"
                ],
                "png": [
                    "summary_plot.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,6a95a24db9848e61173bfd3c05ee2bde"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-08T12:02:42.30178343"
    }
}